modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/leases/LeaseTracker.java

```diff
@@ -239,6 +245,20 @@ public CompletableFuture<ReplicaMeta> awaitPrimaryReplica(
     ) {
         return inBusyLockAsync(busyLock, () -> getOrCreatePrimaryReplicaWaiter(groupId).waitFor(timestamp)
                 .orTimeout(timeout, unit)
+                .thenCompose(replicaMeta -> {
+                    ClusterNode leaseholderNode = clusterNodeResolver.getById(replicaMeta.getLeaseholderId());
+
+                    if (leaseholderNode == null) {
+                        return awaitPrimaryReplica(
+                                groupId,
+                                replicaMeta.getExpirationTime().tick(),
+                                timeout,
```

Comment:
I don't think that it'll  work, it's required to either decrease the timeout for consecutive calls or prepare outer result future with timeout that will be completed explicitly by the inner calls.

Reply:
fixed